api/helpers.go

```diff
@@ -99,31 +99,54 @@ func (a *API) requestAud(ctx context.Context, r *http.Request) string {
 func (a *API) getReferrer(r *http.Request) string {
 	ctx := r.Context()
 	config := a.getConfig(ctx)
-	referrer := ""
-	if reqref := r.Referer(); reqref != "" {
-		base, berr := url.Parse(config.SiteURL)
-		refurl, rerr := url.Parse(reqref)
-		// As long as the referrer came from the site, we will redirect back there
-		if berr == nil && rerr == nil && base.Hostname() == refurl.Hostname() {
-			referrer = reqref
+	reqref := r.Referer()
+	if reqref == "" {
+		return ""
+	}
+
+	base, berr := url.Parse(config.SiteURL)
+	refurl, rerr := url.Parse(reqref)
+	// As long as the referrer came from the site, we will redirect back there
+	if berr == nil && rerr == nil && base.Hostname() == refurl.Hostname() {
+		return reqref
+	}
+
+	// For case when user came from mobile app or other permitted resource - redirect back
+	for _, uri := range config.URIAllowList {
+		puri, perr := url.Parse(uri)
+		if perr == nil && puri.Hostname() == refurl.Hostname() {
```

Comment:
I think here we want to match the **exact** strings, rather than just hostname, the same way google oauth does for redirect URIs